exports/dataset.cpp

```diff
@@ -38,6 +40,49 @@ template <class T> void declare_span(py::module &m, const std::string &suffix) {
   mutable_span_methods<T>::add(span);
 }
 
+template <class... Keys>
+void declare_VariableZipProxy(py::module &m, const std::string &suffix,
+                              const Keys &... keys) {
+  using Proxy = decltype(zip(std::declval<Dataset &>(), keys...));
+  py::class_<Proxy> proxy(m,
+                          (std::string("VariableZipProxy_") + suffix).c_str());
+  proxy.def("__len__", &Proxy::size)
+      .def("__getitem__", &Proxy::operator[])
+      .def("__iter__",
+           [](const Proxy &self) {
+             return py::make_iterator(self.begin(), self.end());
+           },
+           // WARNING The py::keep_alive is really important. It prevents
+           // deletion of the VariableZipProxy when its iterators are still in
+           // use. This is necessary due to the underlying implementation, which
+           // used ranges::view::zip based on temporary gsl::range.
```

Comment:
Interesting, where is the reference counting/monitoring for these temporaries happening?

Reply:
`ranges::view::zip` moves the temporaries into the `zip_view` it is creating. The reference counting for that is done somewhere in `pybind11` I guess.